### PR TITLE
Add initial widget for calculating cell statistics

### LINKED ISSUE #26: Extract cell statistics
Should be possible with Alan's tool and/or `regionprops`

### PR DESCRIPTION
**Pull request by p-j-smith**
Fixes #26 

- added a function `napari_epitools._widgets.cell_statistics_widget`. This creates a widget with:
    - dropdown menus for selecting the input Image and Labels
    - button to calculate statistics
    - button to export statistics to csv file

image

- the following statistics are calculated:
    - area
    - perimeter
    - orientation
    - number of neighbours of each cell

- I'm not sure what units these statistics are in, nor what unit we want them to be in?

- the cell statistics are calculated for every frame in the timeseries

- the stats for the current frame are stored in `Labels.features`. The stats for all frames are stored in `Labels.metadata`, and a callback is used to update `Labels.features` for the current frame when scrolling through the timeseries

- the statistics are currently shown in the status bar when hovering over a cell if the corresponding `Labels` layer is selected - this is done automatically by napari when we set `Labels.features`. However, it would be nice to instead display the statistics in a tooltip, which [should be possible](https://github.com/napari/napari/pull/2658)

- when exporting the stats to csv, the dataframe is saved with a MultiIndex using the frame and the label of each row

### CODE REVIEW

File: src/napari_epitools/_sample_data.py [hunk 1]
```diff
@@ -15,6 +15,8 @@
 
 def load_sample_data():
     """Generates an image"""
-    img_path = Path("sample_data") / "8bitDataset" / "test_image.tif"
+    img_path = (
+        Path("sample_data") / "4d" / "181210_DM_CellMOr_subsub_reg_decon-1_cropped.tif"
+    )
```
Comment (paddyroddy) [inline src/napari_epitools/_sample_data.py:20]: This is the new data they provided?

File: src/napari_epitools/_widget.py [hunk 2]
```diff
@@ -240,3 +247,173 @@ def run() -> tuple[npt.NDArray[np.float64], npt.NDArray[np.int64]]:
...
+) -> None:
+    """Update Labels cell_statistics for current frame"""
+
+    for layer in layers:
+        try:
+            layer.features = layer.metadata["cell_statistics"][frame]
+        except KeyError:
+            pass
+        except IndexError:
+            pass
```
Comment (paddyroddy) [inline src/napari_epitools/_widget.py:344]: Should we log something here?

File: src/napari_epitools/_widget.py [hunk 3]
```diff
@@ -240,3 +247,173 @@ def run() -> tuple[npt.NDArray[np.float64], npt.NDArray[np.int64]]:
...
+        image=image.data,
+        labels=labels.data,
+    )
+
+    # We will use these to update the cell stats at each frame
+    labels.metadata["cell_statistics"] = cell_statistics
+    labels.metadata["graphs"] = graphs
+
+    # confirm calculation finished
+    message = f"'Finished calculating cell statistics for '{labels.name}'"
```
Comment (paddyroddy) [inline src/napari_epitools/_widget.py:363]:
```suggestion
    message = f"Finished calculating cell statistics for '{labels.name}'"
```

File: src/napari_epitools/_widget.py [hunk 4]
```diff
@@ -240,3 +247,173 @@ def run() -> tuple[npt.NDArray[np.float64], npt.NDArray[np.int64]]:
...
+    napari.utils.notifications.show_info(message)
+
+    # Set cell stats for the current frame
+    viewer = napari.current_viewer()
+    current_frame = viewer.dims.current_step[0]
+    try:
+        labels.features = cell_statistics[current_frame]
+    except IndexError:
+        pass
```
Comment (paddyroddy) [inline src/napari_epitools/_widget.py:372]: Log?

File: src/napari_epitools/_widget.py [hunk 5]
```diff
@@ -240,3 +247,173 @@ def run() -> tuple[npt.NDArray[np.float64], npt.NDArray[np.int64]]:
...
+    show_file_dialog = app.get_obj("show_file_dialog")
+    filename = show_file_dialog(
+        mode=FileDialogMode.OPTIONAL_FILE,
+        caption="Specify file to save Epitools cell statistics",
+        start_path=None,
+        filter="*.csv",
+    )
+
+    if filename is None:
+        return
```
Comment (paddyroddy) [inline src/napari_epitools/_widget.py:390]: Log something?

File: src/napari_epitools/_widget.py [hunk 6]
```diff
@@ -240,3 +247,173 @@ def run() -> tuple[npt.NDArray[np.float64], npt.NDArray[np.int64]]:
...
+
+    df = pd.concat(
+        [pd.DataFrame.from_dict(stats).set_index("index") for stats in cell_statistics],
+        keys=list(range(len(cell_statistics))),
+    )
+    df.to_csv(filename, index_label=["frame", "label"])
+
+    # confirm export
+    message = f"'{labels.name}' cell statistics written to {filename}"
```
Comment (paddyroddy) [inline src/napari_epitools/_widget.py:418]: Are the first quotes needed? If so we should maybe do both?
```suggestion
    message = f"'{labels.name}' cell statistics written to '{filename}'"
```

File: src/napari_epitools/analysis/cell_statistics.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    """Create graph of neighbouring cells"""
+
+    graphs = [skimage.graph.RAG(frame_labels) for frame_labels in labels]
+
+    # remove the background if it exists
+    for graph in graphs:
+        try:
+            graph.remove_node(0)
+        except networkx.exception.NetworkXError:
+            pass
```
Comment (paddyroddy) [inline src/napari_epitools/analysis/cell_statistics.py:79]: Log something?

**Review by paddyroddy — APPROVED** [review]
Looks great. As for units, I reckon let's get this in and then we can perfect.

**Comment by p-j-smith** [comment]
> As for units, I reckon let's get this in and then we can perfect.

yep that makes sense

File: src/napari_epitools/_sample_data.py (hunk 1 above)
Comment (p-j-smith) [reply src/napari_epitools/_sample_data.py:20]: yeah that's the new 4d data - there's only 5 frames in the timeseries so perfect for our needs for now

File: src/napari_epitools/_widget.py (hunk 2 above)
Comment (p-j-smith) [reply src/napari_epitools/_widget.py:344]: Shall we set up logging in a separate pr?

**Pull request merged**